ArcGISRuntimeSDKQt_SampleViewers/ArcGISRuntimeSDKQt_Samples/GAnalytics.cpp

```diff
@@ -0,0 +1,177 @@
+#include "GAnalytics.h"
+
+#include <QDebug>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QNetworkAccessManager>
+#include <QNetworkReply>
+#include <QOperatingSystemVersion>
+#include <QQmlEngine>
+#include <QSettings>
+#include <QUuid>
+
+#ifdef GANALYTICS_API_KEY
+#ifdef GANALYTICS_STREAM_ID
+GAnalytics::GAnalytics(QObject* parent /* = nullptr */):
+  QObject(parent),
+  m_apiSecret(QUOTE(GANALYTICS_API_KEY)),
+  m_measurementId(QUOTE(GANALYTICS_STREAM_ID))
+{
+}
+#endif // GANALYTICS_STREAM_ID
+#endif // GANALYTICS_API_KEY
+
+GAnalytics::GAnalytics(QObject* parent /* = nullptr */):
+  QObject(parent),
+  m_apiSecret(""),
+  m_measurementId("")
+{
+}
+
+GAnalytics::~GAnalytics()
+{
+  endSession();
+}
+
+void GAnalytics::init()
+{
+  if (m_apiSecret == "" || m_measurementId == "")
+  {
+    m_telemetryEnabled = false;
+    setIsVisible(false);
+    return;
+  }
+
+  QSettings settings;
+  generateClientId();
+  if (!settings.contains("GAnalytics-telemetry-enabled"))
+  {
+    m_telemetryEnabled = true;
+    settings.setValue("GAnalytics-telemetry-enabled", m_telemetryEnabled);
+    setIsVisible(true);
+  }
+  else
+  {
+    m_telemetryEnabled = settings.value("GAnalytics-telemetry-enabled").toBool();
+  }
+}
+
+void GAnalytics::postEvent(const QString eventName, QVariantMap parameters)
+{
+  if (!m_telemetryEnabled || m_apiSecret == "" || m_measurementId == "")
+    return;
+
+#ifdef CPP_VIEWER
+  parameters.insert("sample_viewer_type", "Cpp");
+#else // QML_VIEWER
+  parameters.insert("sample_viewer_type", "QML");
+#endif // CPP_VIEWER / QML_VIEWER
+
+  auto os = QOperatingSystemVersion::current();
+
+  parameters.insert("operating_system", os.name() + " " + QString::number(os.majorVersion()) + "." + QString::number(os.minorVersion()));
+
+  int sessionTime = (QDateTime::currentSecsSinceEpoch()-m_startTime)*1000;
+
+  // Google will not accept the POST request if the engagement time is 0 or undefined
+  if (sessionTime <= 0)
+    sessionTime = 1;
+
+  parameters.insert("engagement_time_msec", sessionTime);
+
+  QJsonObject event;
+  event.insert("name", eventName);
+  event.insert("params", QJsonObject::fromVariantMap(parameters));
+
+  QJsonObject body;
+  body.insert("client_id", m_clientId);
+  body.insert("events", event);
+  QNetworkAccessManager* mgr = new QNetworkAccessManager();
+  const QUrl url(QString("https://google-analytics.com/mp/collect?"
```

Comment:
🙃 Consider `QString.arg` to make it a bit easier to maintain vs concatenated strings